### PR TITLE
Cube coadding

### PR DESCRIPTION
**Pull request by rcooke-ast**
Mostly just docs, with a few bug fixes and improvements to the cube coadding.

### CODE REVIEW

File: pypeit/core/datacube.py [hunk 1]
```diff
@@ -153,6 +209,13 @@ def make_whitelight(all_ra, all_dec, all_wave, all_sci, all_wghts, all_idx,
         all_ivar (`numpy.ndarray`_, optional):
             Inverse variance of each pixel from all spec2d files. If provided,
             inverse variance images will be calculated and return for each white light image.
+        whitelightWCS (`astropy.wcs.wcs.WCS`_):
+            The WCS of a reference white light image. If supplied, you must also
+            supply numra and numdec.
+        numra (int):
```
Comment (profxj) [inline pypeit/core/datacube.py:215]: optional

File: pypeit/scripts/coadd_datacube.py [hunk 2]
```diff
@@ -319,7 +339,8 @@ def coadd_cube(files, parset, overwrite=False):
     primary_hdu = fits.PrimaryHDU(header=spec2DObj.head0)
```
Comment (profxj) [inline pypeit/scripts/coadd_datacube.py:339]: Consider making this a simple DataContainer.
You could model it after `OneSpec` in `coadd1d.py`

**Review by profxj — APPROVED** [review]
One nit and one potentially substantial suggestion.
Take it as you wish.

File: pypeit/core/datacube.py (hunk 1 above)
Comment (rcooke-ast) [reply pypeit/core/datacube.py]: done

File: pypeit/core/datacube.py [hunk 3]
```diff
@@ -5,15 +5,146 @@
...
+from pypeit import io
+
+
+class DataCube(datamodel.DataContainer):
+    """
+    DataContainer to hold the products of a datacube
+
+    See the datamodel for argument descriptions
+
+    Args:
```
Comment (jhennawi) [inline pypeit/core/datacube.py:31]: This does not look up to date, and shouldn't the args be listed under the __init__.

File: pypeit/core/datacube.py [hunk 4]
```diff
@@ -126,8 +257,63 @@ def dar_correction(wave_arr, coord, obstime, location, pressure, temperature, re
...
+            An integer identifier indicating which spec2d file each pixel originates from.
+            For example, a 0 would indicate that a pixel originates from the first spec2d
+            frame listed in the input file. a 1 would indicate that this pixel originates
+            from the second spec2d file, and so forth.
+        dspat (float):
+            The size of each spaxel on the sky (in degrees)
+        ref_filename (str):
+            A fits filename of a reference image to be used when generating white light
+            images. Note, the fits file must have a valid 3D WCS.
+        numfiles (int, optional):
```
Comment (jhennawi) [inline pypeit/core/datacube.py:287]: Omit redundant arguments?

File: pypeit/core/datacube.py [hunk 5]
```diff
@@ -126,8 +257,63 @@ def dar_correction(wave_arr, coord, obstime, location, pressure, temperature, re
...
 
 
+def make_whitelight_fromref(all_ra, all_dec, all_wave, all_sci, all_wghts, all_idx, dspat,
+                            ref_filename, numfiles=None):
+    """ Generate a whitelight image of every input frame,
+    based on a reference image. Note the, the reference
+    image must have a well-defined WCS.
+
+    Args:
+        all_ra (`numpy.ndarray`_):
```
Comment (jhennawi) [inline pypeit/core/datacube.py:267]: Indicate shapes of all inputs as is our doc convention.

File: pypeit/core/datacube.py [hunk 6]
```diff
@@ -261,13 +463,13 @@ def compute_weights(all_ra, all_dec, all_wave, all_sci, all_ivar, all_wghts, all
             Counts of each pixel from all spec2d files
         all_ivar (`numpy.ndarray`_):
```
Comment (jhennawi) [inline pypeit/core/datacube.py:464]: Document shapes of arrays.

**Review by jhennawi — CHANGES_REQUESTED** [review]
Looks good @rcooke-ast. Minor doc fixes requested.

File: pypeit/core/datacube.py (hunk 5 above)
Comment (rcooke-ast) [reply pypeit/core/datacube.py:267]: Done

File: pypeit/core/datacube.py (hunk 4 above)
Comment (rcooke-ast) [reply pypeit/core/datacube.py:287]: I'm not sure what you mean by redundant, here. `numfiles` is passed to  `make_whitelight`

File: pypeit/core/datacube.py (hunk 6 above)
Comment (rcooke-ast) [reply pypeit/core/datacube.py:464]: Updated

**Review by rcooke-ast — COMMENTED** [review]
Addresses @jhennawi's comments

**Comment by rcooke-ast** [comment]
Dev suite passes:
image

File: pypeit/core/datacube.py (hunk 4 above)
Comment (jhennawi) [reply pypeit/core/datacube.py:294]: It is just the size of all_idx, so just determine if in make_whitelight from that array, unless I'm missing something?

**Review by jhennawi — APPROVED** [review]

**Comment by profxj** [comment]
merging

**Pull request merged**